src/DrCabal/Profile/Stacked.hs

```diff
@@ -150,6 +153,12 @@ formatChart start end width libs = unlines $ concat
     longestPhase :: Word64
     longestPhase = List.maximum $ map (phaseTotal . snd) entries
 
+    totalAllPhases :: Word64
+    totalAllPhases = getSum $ foldMap (Sum . phaseTotal . snd) entries
```

Comment:
`foldMap` doesn't perform the strict evaluation of the result and is not even tail-recursive in this case. Here it's better to use the standard function `sum`. And the code will be slightly shorter and easier to understand as a consequence 🙂 
```suggestion
    totalAllPhases = sum $ map (phaseTotal . snd) entries
```

Reply:
Huh, didn’t know that about `foldMap` — do you have a reference I could look though? I worried that `sum . map f` might traverse the list twice, though then again I suppose list fusion will take care of that, if it’s even a problem here in the first place…

Reply:
`dr-cabal` uses `relude` which implements `sum` using `foldl'`:

* https://hackage.haskell.org/package/relude-1.1.0.0/docs/src/Relude.Foldable.Fold.html#sum

It's the same implementation as in `base` starting from some recent version of GHC.

It won't traverse the list twice because `map` produces output lists lazily and `foldl'` consumes lazily but with a strict accumulator.

List fusion doesn't always reduce the number of traversals. The naive composition of `map`, `filter` and `foldl'` will traverse the list once anyway. List fusion only helps to optimise the performance by rewriting the composition into a single `fold'`-like function. Even though, a composition of `foldl'` and `map` should indeed be optimised with list fusion 🙂

Reply:
> List fusion doesn't always reduce the number of traversals. The naive composition of `map`, `filter` and `foldl'` will traverse the list once anyway. List fusion only helps to optimise the performance by rewriting the composition into a single `fold'`-like function. Even though, a composition of `foldl'` and `map` should indeed be optimised with list fusion 🙂

Ah, I see — thanks for explaining! This makes perfect sense. (Laziness is the one part of Haskell I still have some trouble reasoning about…)

> `foldl'` consumes lazily but with a strict accumulator.

Hmm, so would this be why `foldMap` is bad — because it accumulates lazily instead? Or am I getting confused again?

Reply:
The naive implementation of `foldMap` is:

```haskell
foldMap f = foldr (mappend . f) mempty
```

If we expand the definition of `foldr`, we'll get:

```haskell
foldMap f [] = mempty
foldMap f (x : xs) = f x <> foldMap f xs
```

And if use the `Sum` constructor, we'll get something like:

```haskell
add [] = 0
add (x : xs) = x + add xs
```

So, this is just not tail-recursive and will keep the entire evaluation stack on memory instead of consuming the list on-demand and immediately calculating the accumulator.

Not a problem with lazy evaluation here, just general algorithmic deficiency 🙂

Reply:
Oh, I think I see… thanks for taking the time to explain this!